server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -881,7 +878,16 @@ private static void parseCopy(String field, ParseContext context) throws IOExcep
     }
 
     // looks up a child mapper, but takes into account field names that expand to objects
-    private static Mapper getMapper(ObjectMapper objectMapper, String fieldName, String[] subfields) {
+    private static Mapper getMapper(final ParseContext context, ObjectMapper objectMapper, String fieldName, String[] subfields) {
+        String fieldPath = context.path().pathAsText(fieldName);
+        if (context.mapperService().isMetadataField(fieldPath)) {
+            for (MetadataFieldMapper metadataFieldMapper : context.docMapper().mapping().getMetadataMappers()) {
```

Comment:
I just noticed we're doing a linear scan through all the metadata mappers. Maybe we should make sure to store them as a map from name -> mapper to avoid this overhead?

Reply:
You are right. I had this point in mind for improvement. I just saw that linear scan happens at: 
https://github.com/elastic/elasticsearch/blob/b8b3a9d4329ccadfa7332922047d6f205c0d7ccb/server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java#L101

I understand that pre/postParse() `internalParseDocument()` is called once per document, while `getMapper()` is called once per field. 

I will fix this asap!

Reply:
Fixed this last bit.  Can please you have one last look?

Thank you for patience and excellent guidance.

Reply:
The change makes sense to me.

It feels redundant that we store an array of metadata mappers, a map from class -> mapper, and a map from name -> mapper. This could be simplified, but we don't necessarily need to do it in this PR.